File: C#/WindowsPhone/LGame-XNA-lib/Loon.Action.Map/TileMap.cs
using Loon.Core;
using Loon.Action.Sprite;
using Loon.Core.Graphics.Opengl;
using System.Collections.Generic;
using Loon.Core.Geom;
using Loon.Utils;
using System;
using Loon.Core.Graphics;
namespace Loon.Action.Map {
	
	public class TileMap : LObject, ISprite, LRelease {
	
		public interface DrawListener {
	
			void Update(long elapsedTime);
	
			void Draw(GLEx g, float x, float y);
	
		}
	
		public class Tile {
	
			internal int id;
	
			internal int imgId;
	
			public Attribute attribute;
	
			internal bool isAnimation;
	
			public Animation animation;
	
		}
	
		private int firstTileX;
	
		private int firstTileY;
	
		private int lastTileX;
	
		private int lastTileY;
	
		public DrawListener listener;
	
		private LTexturePack imgPack;

        private List<TileMap.Tile> arrays = new List<TileMap.Tile>(10);

        private List<Animation> animations = new List<Animation>();
	
		private readonly int maxWidth, maxHeight;
	
		private readonly Field2D field;
	
		private float lastOffsetX, lastOffsetY;
	
		private Vector2f offset;
	
		private Loon.Core.Graphics.Opengl.LTexture.Format format;
	
		private bool active, dirty;
	
		private bool visible;
	
		private bool playAnimation;
	
		public TileMap(string fileName, int tileWidth, int tileHeight) :this(fileName, tileWidth, tileHeight, LSystem.screenRect.width,
					LSystem.screenRect.height, Loon.Core.Graphics.Opengl.LTexture.Format.LINEAR){
		
		}
	
		public TileMap(string fileName, int tileWidth, int tileHeight, int mWidth,
				int mHeight, Loon.Core.Graphics.Opengl.LTexture.Format format):	this(TileMapConfig.LoadAthwartArray(fileName), tileWidth, tileHeight,
					mWidth, mHeight, format) {
		
		}
	
		public TileMap(int[][] maps, int tileWidth, int tileHeight, int mWidth,
				int mHeight, Loon.Core.Graphics.Opengl.LTexture.Format format):this(new Field2D(maps, tileWidth, tileHeight), mWidth, mHeight, format) {
			
		}
	
		public TileMap(Field2D field):this(field, LSystem.screenRect.width, LSystem.screenRect.height,
					Loon.Core.Graphics.Opengl.LTexture.Format.LINEAR) {
			
		}
	
		public TileMap(Field2D field, Loon.Core.Graphics.Opengl.LTexture.Format format):this(field, LSystem.screenRect.width, LSystem.screenRect.height, format) {
			
		}
	
		public TileMap(Field2D field, int mWidth, int mHeight, Loon.Core.Graphics.Opengl.LTexture.Format format) {
			this.field = field;
			this.maxWidth = mWidth;
			this.maxHeight = mHeight;
			this.offset = new Vector2f(0, 0);
			this.imgPack = new LTexturePack();
			this.format = format;
			this.lastOffsetX = -1;
			this.lastOffsetY = -1;
			this.active = true;
			this.dirty = true;
			this.visible = true;
			imgPack.SetFormat(format);
		}
	
		public static TileMap LoadCharsMap(string resName, int tileWidth,
				int tileHeight) {
			return new TileMap(TileMapConfig.LoadCharsField(resName, tileWidth,
					tileHeight));
		}
	
		public void SetImagePack(string file) {
			if (imgPack != null) {
				imgPack.Dispose();
				imgPack = null;
			}
			this.active = false;
			this.dirty = true;
			imgPack = new LTexturePack(file);
			imgPack.Packed(format);
		}
	
		public void RemoveTile(int id) {
			foreach (Tile tile  in  arrays) {
				if (tile.id == id) {
					if (tile.isAnimation) {
						CollectionUtils.Remove(animations,tile.animation);
					}
                    CollectionUtils.Remove(arrays,tile);
				}
			}
			if (animations.Count == 0) {
				playAnimation = false;
			}
		}
	
		public int PutAnimationTile(int id, Animation animation, Attribute attribute) {
			if (active) {
				TileMap.Tile tile = new TileMap.Tile();
				tile.id = id;
				tile.imgId = -1;
				tile.attribute = attribute;
				if (animation != null && animation.GetTotalFrames() > 0) {
					tile.isAnimation = true;
					tile.animation = animation;
					playAnimation = true;
				}
				CollectionUtils.Add(animations,animation);
                CollectionUtils.Add(arrays,tile);
				dirty = true;
				return tile.imgId;
			} else {
				throw new Exception(
						"Map is no longer active, you can not add new tiles !");
			}
		}
	
		public int PutAnimationTile(int id, string res, int w, int h, int timer) {
			return PutAnimationTile(id,
					Animation.GetDefaultAnimation(res, w, h, timer), null);
		}
	
		public int PutAnimationTile(int id, Animation animation) {
			return PutAnimationTile(id, animation, null);
		}
	
		public int PutTile(int id, LImage img, Attribute attribute) {
			if (active) {
				TileMap.Tile tile = new TileMap.Tile();
				tile.id = id;
				tile.imgId = imgPack.PutImage(img);
				tile.attribute = attribute;
				CollectionUtils.Add(arrays,tile);
				dirty = true;
				return tile.imgId;
			} else {
				throw new Exception(
						"Map is no longer active, you can not add new tiles !");
			}
		}
	
		public int PutTile(int id, LImage img) {
			return PutTile(id, img, null);
		}

        public int PutTile(int id, LTexture img, Attribute attribute)
        {
            if (active)
            {
                TileMap.Tile tile = new TileMap.Tile();
                tile.id = id;
                tile.imgId = imgPack.PutImage(img);
                tile.attribute = attribute;
                CollectionUtils.Add(arrays, tile);
                dirty = true;
                return tile.imgId;
            }
            else
            {
                throw new Exception(
                        "Map is no longer active, you can not add new tiles !");
            }
        }
	
		public int PutTile(int id, LTexture img) {
			return PutTile(id, img, null);
		}
	
		public int PutTile(int id, string res, Attribute attribute) {
			if (active) {
				TileMap.Tile tile = new TileMap.Tile();
				tile.id = id;
				tile.imgId = imgPack.PutImage(res);
				tile.attribute = attribute;
				CollectionUtils.Add(arrays,tile);
				dirty = true;
				return tile.imgId;
			} else {
				throw new Exception(
						"Map is no longer active, you can not add new tiles !");
			}
		}
	
		public int PutTile(int id, string res) {
			return PutTile(id, res, null);
		}

        public void PutTile(int id, int imgId, Attribute attribute)
        {
            if (active)
            {
                TileMap.Tile tile = new TileMap.Tile();
                tile.id = id;
                tile.imgId = imgId;
                tile.attribute = attribute;
                CollectionUtils.Add(arrays, tile);
                dirty = true;
            }
            else
            {
                throw new Exception(
                        "Map is no longer active, you can not add new tiles !");
            }
        }
	
		public void PutTile(int id, int imgId) {
			PutTile(id, imgId, null);
		}
	
		public TileMap.Tile GetTile(int id) {
			foreach (Tile tile  in  arrays) {
				if (tile.id == id) {
					return tile;
				}
			}
			return null;
		}
	
		public int[][] GetMap() {
			return field.GetMap();
		}
	
		public bool IsActive() {
			return active;
		}
	
		public void Completed() {
			if (imgPack != null) {
				imgPack.Packed(format);
				active = false;
			}
		}
	
		public Loon.Core.Graphics.Opengl.LTexture.Format GetFormat() {
			return format;
		}
	
		public int GetTileID(int x, int y) {
			if (x >= 0 && x < field.GetWidth() && y >= 0 && y < field.GetHeight()) {
				return field.GetType(y, x);
			} else {
				return -1;
			}
		}
	
		public void SetTileID(int x, int y, int id) {
			if (x >= 0 && x < field.GetWidth() && y >= 0 && y < field.GetHeight()) {
				field.SetType(y, x, id);
			}
		}
	
		public void Draw(GLEx g) {
			Draw(g, null, offset.X(), offset.Y());
		}
	
		public void Draw(GLEx g, SpriteBatch batch, float offsetX, float offsetY) {
			bool useBatch = (batch != null);
			if (!dirty && lastOffsetX == offsetX && lastOffsetY == offsetY) {
				imgPack.GLCache();
				if (playAnimation) {
					int[][] maps = field.GetMap();
					for (int i = firstTileX; i < lastTileX; i++) {
						for (int j = firstTileY; j < lastTileY; j++) {
							if (i > -1 && j > -1 && i < field.width
									&& j < field.height) {
								int id = maps[j][i];
								foreach (Tile tile  in  arrays) {
									if (tile.isAnimation && tile.id == id) {
										if (useBatch) {
											batch.Draw(
													tile.animation.GetSpriteImage(),
													field.TilesToWidthPixels(i)
															+ offsetX,
													field.TilesToHeightPixels(j)
															+ offsetY,
													field.GetTileWidth(),
													field.GetTileHeight());
										} else {
											g.DrawTexture(
													tile.animation.GetSpriteImage(),
													field.TilesToWidthPixels(i)
															+ offsetX,
													field.TilesToHeightPixels(j)
															+ offsetY,
													field.GetTileWidth(),
													field.GetTileHeight());
										}
									}
								}
							}
						}
					}
				}
			} else {
				if (arrays.Count == 0) {
					throw new Exception("Not to add any tiles !");
				}
				imgPack.GLBegin();
				firstTileX = field.PixelsToTilesWidth(-offsetX);
				firstTileY = field.PixelsToTilesHeight(-offsetY);
				lastTileX = firstTileX + field.PixelsToTilesWidth(maxWidth) + 1;
				lastTileX = MathUtils.Min(lastTileX, field.width);
				lastTileY = firstTileY + field.PixelsToTilesHeight(maxHeight) + 1;
				lastTileY = MathUtils.Min(lastTileY, field.height);
				int[][] maps = field.GetMap();
				for (int i = firstTileX; i < lastTileX; i++) {
					for (int j = firstTileY; j < lastTileY; j++) {
						if (i > -1 && j > -1 && i < field.width && j < field.height) {
							int id = maps[j][i];
							foreach (Tile tile  in  arrays) {
								if (playAnimation) {
									if (tile.id == id) {
										if (tile.isAnimation) {
											if (useBatch) {
												batch.Draw(
														tile.animation
																.GetSpriteImage(),
														field.TilesToWidthPixels(i)
																+ offsetX,
														field.TilesToHeightPixels(j)
																+ offsetY, field
																.GetTileWidth(),
														field.GetTileHeight());
											} else {
												g.DrawTexture(
														tile.animation
																.GetSpriteImage(),
														field.TilesToWidthPixels(i)
																+ offsetX,
														field.TilesToHeightPixels(j)
																+ offsetY, field
																.GetTileWidth(),
														field.GetTileHeight());
											}
										} else {
											imgPack.Draw(tile.imgId,
													field.TilesToWidthPixels(i)
															+ offsetX,
													field.TilesToHeightPixels(j)
															+ offsetY,
													field.GetTileWidth(),
													field.GetTileHeight());
										}
									}
								} else if (tile.id == id) {
									imgPack.Draw(tile.imgId,
											field.TilesToWidthPixels(i) + offsetX,
											field.TilesToHeightPixels(j) + offsetY,
											field.GetTileWidth(),
											field.GetTileHeight());
								}
	
							}
						}
					}
				}
				imgPack.GLEnd();
				lastOffsetX = offsetX;
				lastOffsetY = offsetY;
				dirty = false;
			}
			if (listener != null) {
				listener.Draw(g, offsetX, offsetY);
			}
		}
	
		public int[] GetLimit() {
			return field.GetLimit();
		}
	
		public void SetLimit(int[] limit) {
			field.SetLimit(limit);
		}
	
		public bool IsHit(int px, int py) {
			return field.IsHit(px, py);
		}
	
		public bool IsHit(Vector2f v) {
			return field.IsHit(v);
		}
	
		public Vector2f GetTileCollision(LObject o, float newX, float newY) {
			newX = MathUtils.Ceil(newX);
			newY = MathUtils.Ceil(newY);
	
			float fromX = MathUtils.Min(o.GetX(), newX);
			float fromY = MathUtils.Min(o.GetY(), newY);
			float toX = MathUtils.Max(o.GetX(), newX);
			float toY = MathUtils.Max(o.GetY(), newY);
	
			int fromTileX = field.PixelsToTilesWidth(fromX);
			int fromTileY = field.PixelsToTilesHeight(fromY);
			int toTileX = field.PixelsToTilesWidth(toX + o.GetWidth() - 1f);
			int toTileY = field.PixelsToTilesHeight(toY + o.GetHeight() - 1f);
	
			for (int x = fromTileX; x <= toTileX; x++) {
				for (int y = fromTileY; y <= toTileY; y++) {
					if ((x < 0) || (x >= field.GetWidth())) {
						return new Vector2f(x, y);
					}
					if ((y < 0) || (y >= field.GetHeight())) {
						return new Vector2f(x, y);
					}
					if (!this.IsHit(x, y)) {
						return new Vector2f(x, y);
					}
				}
			}
	
			return null;
		}
	
		public int GetTileIDFromPixels(Vector2f v) {
			return GetTileIDFromPixels(v.x, v.y);
		}
	
		public int GetTileIDFromPixels(float sx, float sy) {
			float x = (sx + offset.GetX());
			float y = (sy + offset.GetY());
			Vector2f tileCoordinates = pixelsToTiles(x, y);
			return GetTileID(MathUtils.Round(tileCoordinates.GetX()),
					MathUtils.Round(tileCoordinates.GetY()));
		}
	
		public Vector2f pixelsToTiles(float x, float y) {
			float xprime = x / field.GetTileWidth() - 1;
			float yprime = y / field.GetTileHeight() - 1;
			return new Vector2f(xprime, yprime);
		}
	
		public Field2D GetField() {
			return field;
		}
	
		public int TilesToPixelsX(float x) {
			return field.TilesToWidthPixels(x);
		}
	
		public int TilesToPixelsY(float y) {
			return field.TilesToHeightPixels(y);
		}
	
		public int PixelsToTilesWidth(float x) {
			return field.PixelsToTilesWidth(x);
		}
	
		public int PixelsToTilesHeight(float y) {
			return field.PixelsToTilesHeight(y);
		}
	
		public Vector2f TilesToPixels(float x, float y) {
			float xprime = x * field.GetTileWidth() - offset.GetX();
			float yprime = y * field.GetTileHeight() - offset.GetY();
			return new Vector2f(xprime, yprime);
		}
	
		public void SetOffset(float x, float y) {
			this.offset.Set(x, y);
		}
	
		public void SetOffset(Vector2f offset) {
			this.offset.Set(offset);
		}
	
		public Vector2f GetOffset() {
			return offset;
		}
	
		public int GetTileWidth() {
			return field.GetTileWidth();
		}
	
		public int GetTileHeight() {
			return field.GetTileHeight();
		}
	
		public override int GetHeight() {
			return field.GetHeight() * field.GetTileWidth();
		}
	
		public override int GetWidth() {
			return field.GetWidth() * field.GetTileHeight();
		}
	
		public int GetRow() {
			return field.GetWidth();
		}
	
		public int GetCol() {
			return field.GetHeight();
		}
	
		public DrawListener GetListener() {
			return listener;
		}
	
		public void SetListener(DrawListener listener) {
			this.listener = listener;
		}
	
		public bool IsDirty() {
			return dirty;
		}
	
		public void SetDirty(bool dirty) {
			this.dirty = dirty;
		}
	
		public void SetVisible(bool v) {
			this.visible = v;
		}
	
		public bool IsVisible() {
			return visible;
		}
	
		public void CreateUI(GLEx g) {
			if (!visible) {
				return;
			}
			if (GetX() != 0 || GetY() != 0) {
				g.Translate(GetX(), GetY());
			}
			Draw(g);
			if (GetX() != 0 || GetY() != 0) {
				g.Translate(-GetX(), -GetY());
			}
		}
	
		public RectBox GetCollisionBox() {
			return GetRect(X() + offset.x, Y() + offset.y, field.GetTileWidth()
					* field.GetWidth(), field.GetTileHeight() * field.GetHeight());
		}
	
		public LTexture GetBitmap() {
			return imgPack.GetTexture();
		}
	
		public override void Update(long elapsedTime) {
			if (playAnimation && animations.Count > 0) {
				foreach (Animation a  in  animations) {
					a.Update(elapsedTime);
				}
			}
			if (listener != null) {
				listener.Update(elapsedTime);
			}
		}
	
		public void StartAnimation() {
			playAnimation = true;
		}
	
		public void StopAnimation() {
			playAnimation = false;
		}
	
		public virtual void Dispose() {
			visible = false;
			playAnimation = false;
			animations.Clear();
			if (imgPack != null) {
				imgPack.Dispose();
			}
		}
	
	}
}

File: Loon-0.5-Alpha/Loon-Neo/src/loon/action/map/TileMap.java
package loon.action.map;

import java.io.IOException;

import loon.LObject;
import loon.LRelease;
import loon.LSystem;
import loon.LTexture;
import loon.LTexture.Format;
import loon.action.sprite.Animation;
import loon.action.sprite.ISprite;
import loon.action.sprite.SpriteBatch;
import loon.canvas.Image;
import loon.geom.RectBox;
import loon.geom.Vector2f;
import loon.opengl.GLEx;
import loon.opengl.LTexturePack;
import loon.utils.MathUtils;
import loon.utils.TArray;

public class TileMap extends LObject implements ISprite, LRelease {

	/**
	 * 
	 */
	private static final long serialVersionUID = 2419037580406911982L;

	public static interface DrawListener {

		public void update(long elapsedTime);

		public void draw(GLEx g, float x, float y);

	}

	public static class Tile {

		int id;

		int imgId;

		public Attribute attribute;

		boolean isAnimation;

		public Animation animation;

	}

	private int firstTileX;

	private int firstTileY;

	private int lastTileX;

	private int lastTileY;

	public DrawListener listener;

	private LTexturePack imgPack;

	private TArray<TileMap.Tile> arrays = new TArray<TileMap.Tile>(10);

	private TArray<Animation> animations = new TArray<Animation>();

	private final int maxWidth, maxHeight;

	private final Field2D field;

	private int lastOffsetX, lastOffsetY;

	private Vector2f offset;

	private Format format;

	private boolean active, dirty;

	private boolean visible;

	private boolean playAnimation;

	public TileMap(String fileName, int tileWidth, int tileHeight)
			throws IOException {
		this(fileName, tileWidth, tileHeight, LSystem.viewSize.getWidth(),
				LSystem.viewSize.getHeight(), Format.LINEAR);
	}

	public TileMap(String fileName, int tileWidth, int tileHeight, int mWidth,
			int mHeight, Format format) throws IOException {
		this(TileMapConfig.loadAthwartArray(fileName), tileWidth, tileHeight,
				mWidth, mHeight, format);
	}

	public TileMap(int[][] maps, int tileWidth, int tileHeight, int mWidth,
			int mHeight, Format format) {
		this(new Field2D(maps, tileWidth, tileHeight), mWidth, mHeight, format);
	}

	public TileMap(Field2D field) {
		this(field, LSystem.viewSize.getWidth(),
				LSystem.viewSize.getHeight(),
				Format.LINEAR);
	}

	public TileMap(Field2D field, Format format) {
		this(field, LSystem.viewSize.getWidth(),
				LSystem.viewSize.getHeight(), format);
	}

	public TileMap(Field2D field, int mWidth, int mHeight, Format format) {
		this.field = field;
		this.maxWidth = mWidth;
		this.maxHeight = mHeight;
		this.offset = new Vector2f(0, 0);
		this.imgPack = new LTexturePack();
		this.format = format;
		this.lastOffsetX = -1;
		this.lastOffsetY = -1;
		this.active = true;
		this.dirty = true;
		this.visible = true;
		imgPack.setFormat(format);
	}

	public static TileMap loadCharsMap(String resName, int tileWidth,
			int tileHeight) {
		return new TileMap(TileMapConfig.loadCharsField(resName, tileWidth,
				tileHeight));
	}

	public void setImagePack(String file) {
		if (imgPack != null) {
			imgPack.close();
			imgPack = null;
		}
		this.active = false;
		this.dirty = true;
		imgPack = new LTexturePack(file);
		imgPack.packed(format);
	}

	public void removeTile(int id) {
		for (Tile tile : arrays) {
			if (tile.id == id) {
				if (tile.isAnimation) {
					animations.remove(tile.animation);
				}
				arrays.remove(tile);
			}
		}
		if (animations.size == 0) {
			playAnimation = false;
		}
	}

	public int putAnimationTile(int id, Animation animation, Attribute attribute) {
		if (active) {
			TileMap.Tile tile = new TileMap.Tile();
			tile.id = id;
			tile.imgId = -1;
			tile.attribute = attribute;
			if (animation != null && animation.getTotalFrames() > 0) {
				tile.isAnimation = true;
				tile.animation = animation;
				playAnimation = true;
			}
			animations.add(animation);
			arrays.add(tile);
			dirty = true;
			return tile.imgId;
		} else {
			throw new RuntimeException(
					"Map is no longer active, you can not add new tiles !");
		}
	}

	public int putAnimationTile(int id, String res, int w, int h, int timer) {
		return putAnimationTile(id,
				Animation.getDefaultAnimation(res, w, h, timer), null);
	}

	public int putAnimationTile(int id, Animation animation) {
		return putAnimationTile(id, animation, null);
	}

	public int putTile(int id, Image img, Attribute attribute) {
		if (active) {
			TileMap.Tile tile = new TileMap.Tile();
			tile.id = id;
			tile.imgId = imgPack.putImage(img);
			tile.attribute = attribute;
			arrays.add(tile);
			dirty = true;
			return tile.imgId;
		} else {
			throw new RuntimeException(
					"Map is no longer active, you can not add new tiles !");
		}
	}

	public int putTile(int id, Image img) {
		return putTile(id, img, null);
	}

	public int putTile(int id, LTexture img, Attribute attribute) {
		if (active) {
			TileMap.Tile tile = new TileMap.Tile();
			tile.id = id;
			tile.imgId = imgPack.putImage(img);
			tile.attribute = attribute;
			arrays.add(tile);
			dirty = true;
			return tile.imgId;
		} else {
			throw new RuntimeException(
					"Map is no longer active, you can not add new tiles !");
		}
	}

	public int putTile(int id, LTexture img) {
		return putTile(id, img, null);
	}

	public int putTile(int id, String res, Attribute attribute) {
		if (active) {
			TileMap.Tile tile = new TileMap.Tile();
			tile.id = id;
			tile.imgId = imgPack.putImage(res);
			tile.attribute = attribute;
			arrays.add(tile);
			dirty = true;
			return tile.imgId;
		} else {
			throw new RuntimeException(
					"Map is no longer active, you can not add new tiles !");
		}
	}

	public int putTile(int id, String res) {
		return putTile(id, res, null);
	}

	public void putTile(int id, int imgId, Attribute attribute) {
		if (active) {
			TileMap.Tile tile = new TileMap.Tile();
			tile.id = id;
			tile.imgId = imgId;
			tile.attribute = attribute;
			arrays.add(tile);
			dirty = true;
		} else {
			throw new RuntimeException(
					"Map is no longer active, you can not add new tiles !");
		}
	}

	public void putTile(int id, int imgId) {
		putTile(id, imgId, null);
	}

	public TileMap.Tile getTile(int id) {
		for (Tile tile : arrays) {
			if (tile.id == id) {
				return tile;
			}
		}
		return null;
	}

	public int[][] getMap() {
		return field.getMap();
	}

	public boolean isActive() {
		return active;
	}

	public void completed() {
		if (imgPack != null) {
			imgPack.packed(format);
			active = false;
		}
	}

	public Format getFormat() {
		return format;
	}

	public int getTileID(int x, int y) {
		if (x >= 0 && x < field.getWidth() && y >= 0 && y < field.getHeight()) {
			return field.getType(y, x);
		} else {
			return -1;
		}
	}

	public void setTileID(int x, int y, int id) {
		if (x >= 0 && x < field.getWidth() && y >= 0 && y < field.getHeight()) {
			field.setType(y, x, id);
		}
	}

	public void draw(GLEx g) {
		draw(g, null, offset.x(), offset.y());
	}

	public void draw(GLEx g, SpriteBatch batch, int offsetX, int offsetY) {
		final boolean useBatch = (batch != null);
		if (!dirty && lastOffsetX == offsetX && lastOffsetY == offsetY) {
			imgPack.postCache();
			if (playAnimation) {
				int[][] maps = field.getMap();
				for (int i = firstTileX; i < lastTileX; i++) {
					for (int j = firstTileY; j < lastTileY; j++) {
						if (i > -1 && j > -1 && i < field.getWidth()
								&& j < field.getHeight()) {
							int id = maps[j][i];
							for (Tile tile : arrays) {
								if (tile.isAnimation && tile.id == id) {
									if (useBatch) {
										batch.draw(
												tile.animation.getSpriteImage(),
												field.tilesToWidthPixels(i)
														+ offsetX,
												field.tilesToHeightPixels(j)
														+ offsetY,
												field.getTileWidth(),
												field.getTileHeight());
									} else {
										g.draw(
												tile.animation.getSpriteImage(),
												field.tilesToWidthPixels(i)
														+ offsetX,
												field.tilesToHeightPixels(j)
														+ offsetY,
												field.getTileWidth(),
												field.getTileHeight());
									}
								}
							}
						}
					}
				}
			}
		} else {
			if (arrays.size == 0) {
				throw new RuntimeException("Not to add any tiles !");
			}
		
			imgPack.glBegin();
		
			firstTileX = field.pixelsToTilesWidth(-offsetX);
			firstTileY = field.pixelsToTilesHeight(-offsetY);

			lastTileX = firstTileX + field.pixelsToTilesWidth(maxWidth) + 1;
			lastTileX = MathUtils.min(lastTileX, field.getWidth());
			lastTileY = firstTileY + field.pixelsToTilesHeight(maxHeight) + 1;
			lastTileY = MathUtils.min(lastTileY, field.getHeight());
			int[][] maps = field.getMap();
			for (int i = firstTileX; i < lastTileX; i++) {
				for (int j = firstTileY; j < lastTileY; j++) {
					if (i > -1 && j > -1 && i < field.getWidth()
							&& j < field.getHeight()) {
						int id = maps[j][i];
						for (Tile tile : arrays) {
							if (playAnimation) {
								if (tile.id == id) {
									if (tile.isAnimation) {
										if (useBatch) {
											batch.draw(
													tile.animation
															.getSpriteImage(),
													field.tilesToWidthPixels(i)
															+ offsetX,
													field.tilesToHeightPixels(j)
															+ offsetY, field
															.getTileWidth(),
													field.getTileHeight());
										} else {
											g.draw(
													tile.animation
															.getSpriteImage(),
													field.tilesToWidthPixels(i)
															+ offsetX,
													field.tilesToHeightPixels(j)
															+ offsetY, field
															.getTileWidth(),
													field.getTileHeight());
										}
									} else {
										imgPack.draw(tile.imgId,
												field.tilesToWidthPixels(i)
														+ offsetX,
												field.tilesToHeightPixels(j)
														+ offsetY,
												field.getTileWidth(),
												field.getTileHeight());
									}
								}
							} else if (tile.id == id) {
								imgPack.draw(tile.imgId,
										field.tilesToWidthPixels(i) + offsetX,
										field.tilesToHeightPixels(j) + offsetY,
										field.getTileWidth(),
										field.getTileHeight());
							}

						}
					}
				}
			}
			imgPack.glEnd();
			imgPack.saveCache();
			lastOffsetX = offsetX;
			lastOffsetY = offsetY;
			dirty = false;
		}
		
		if (listener != null) {
			listener.draw(g, offsetX, offsetY);
		}
	}

	public int[] getLimit() {
		return field.getLimit();
	}

	public void setLimit(int[] limit) {
		field.setLimit(limit);
	}

	public boolean isHit(int px, int py) {
		return field.isHit(px, py);
	}

	public boolean isHit(Vector2f v) {
		return field.isHit(v);
	}

	public Vector2f getTileCollision(LObject o, float newX, float newY) {
		newX = MathUtils.ceil(newX);
		newY = MathUtils.ceil(newY);

		float fromX = MathUtils.min(o.getX(), newX);
		float fromY = MathUtils.min(o.getY(), newY);
		float toX = MathUtils.max(o.getX(), newX);
		float toY = MathUtils.max(o.getY(), newY);

		int fromTileX = field.pixelsToTilesWidth(fromX);
		int fromTileY = field.pixelsToTilesHeight(fromY);
		int toTileX = field.pixelsToTilesWidth(toX + o.getWidth() - 1f);
		int toTileY = field.pixelsToTilesHeight(toY + o.getHeight() - 1f);

		for (int x = fromTileX; x <= toTileX; x++) {
			for (int y = fromTileY; y <= toTileY; y++) {
				if ((x < 0) || (x >= field.getWidth())) {
					return new Vector2f(x, y);
				}
				if ((y < 0) || (y >= field.getHeight())) {
					return new Vector2f(x, y);
				}
				if (!this.isHit(x, y)) {
					return new Vector2f(x, y);
				}
			}
		}

		return null;
	}

	public int getTileIDFromPixels(Vector2f v) {
		return getTileIDFromPixels(v.x, v.y);
	}

	public int getTileIDFromPixels(float sx, float sy) {
		float x = (sx + offset.getX());
		float y = (sy + offset.getY());
		Vector2f tileCoordinates = pixelsToTiles(x, y);
		return getTileID(MathUtils.round(tileCoordinates.getX()),
				MathUtils.round(tileCoordinates.getY()));
	}

	public Vector2f pixelsToTiles(float x, float y) {
		float xprime = x / field.getTileWidth() - 1;
		float yprime = y / field.getTileHeight() - 1;
		return new Vector2f(xprime, yprime);
	}

	public Field2D getField() {
		return field;
	}

	public int tilesToPixelsX(float x) {
		return field.tilesToWidthPixels(x);
	}

	public int tilesToPixelsY(float y) {
		return field.tilesToHeightPixels(y);
	}

	public int pixelsToTilesWidth(float x) {
		return field.pixelsToTilesWidth(x);
	}

	public int pixelsToTilesHeight(float y) {
		return field.pixelsToTilesHeight(y);
	}

	/**
	 * 转换坐标为像素坐标
	 * 
	 * @param x
	 * @param y
	 * @return
	 */
	public Vector2f tilesToPixels(float x, float y) {
		float xprime = x * field.getTileWidth() - offset.getX();
		float yprime = y * field.getTileHeight() - offset.getY();
		return new Vector2f(xprime, yprime);
	}

	/**
	 * 设置瓦片位置
	 * 
	 * @param x
	 * @param y
	 */
	public void setOffset(float x, float y) {
		this.offset.set(x, y);
	}

	/**
	 * 设定偏移量
	 * 
	 * @param offset
	 */
	public void setOffset(Vector2f offset) {
		this.offset.set(offset);
	}

	/**
	 * 获得瓦片位置
	 * 
	 * @return
	 */
	public Vector2f getOffset() {
		return offset;
	}

	public int getTileWidth() {
		return field.getTileWidth();
	}

	public int getTileHeight() {
		return field.getTileHeight();
	}

	public int getHeight() {
		return field.getHeight() * field.getTileWidth();
	}

	public int getWidth() {
		return field.getWidth() * field.getTileHeight();
	}

	public int getRow() {
		return field.getWidth();
	}

	public int getCol() {
		return field.getHeight();
	}

	public DrawListener getListener() {
		return listener;
	}

	public void setListener(DrawListener listener) {
		this.listener = listener;
	}

	public boolean isDirty() {
		return dirty;
	}

	public void setDirty(boolean dirty) {
		this.dirty = dirty;
	}

	public void setVisible(boolean v) {
		this.visible = v;
	}

	public boolean isVisible() {
		return visible;
	}

	public void createUI(GLEx g) {
		if (!visible) {
			return;
		}
		if (getX() != 0 || getY() != 0) {
			g.translate(getX(), getY());
		}
		draw(g);
		if (getX() != 0 || getY() != 0) {
			g.translate(-getX(), -getY());
		}
	}

	public RectBox getCollisionBox() {
		return getRect(x() + offset.x, y() + offset.y, field.getTileWidth()
				* field.getWidth(), field.getTileHeight() * field.getHeight());
	}

	public LTexture getBitmap() {
		return imgPack.getTexture();
	}

	public void update(long elapsedTime) {
		if (playAnimation && animations.size > 0) {
			for (Animation a : animations) {
				a.update(elapsedTime);
			}
		}
		if (listener != null) {
			listener.update(elapsedTime);
		}
	}

	public void startAnimation() {
		playAnimation = true;
	}

	public void stopAnimation() {
		playAnimation = false;
	}

	public void close() {
		visible = false;
		playAnimation = false;
		animations.clear();
		if (imgPack != null) {
			imgPack.close();
		}
	}

}
